service/history/hsm/tree.go

```diff
@@ -700,10 +761,116 @@ func (n *Node) root() *Node {
 	return root
 }
 
+// compact filters the operation log based on deletion status. For any operation path:
+// - If any parent in the path is deleted, the operation is excluded
+// - If the target of the operation is deleted, only its DeleteOperation is kept
+// - Otherwise, the operation is included
+func (ol OperationLog) compact() OperationLog {
+	if len(ol) == 0 {
+		return ol
+	}
+
+	root := newOpNode(Key{})
+	for _, op := range ol {
+		node := root.getOrCreateNode(op.Path())
+		if _, ok := op.(DeleteOperation); ok {
+			node.isDeleted = true
+		}
+	}
+
+	return root.collect(ol)
+}
+
+// getOrCreateNode traverses/creates path and returns the final node
+func (n *opNode) getOrCreateNode(path []Key) *opNode {
+	current := n
+	for _, key := range path {
+		next, exists := current.children[key]
+		if !exists {
+			next = newOpNode(key)
+			current.children[key] = next
+		}
+		current = next
+	}
+	return current
+}
+
 func isPathPrefix(prefix, path []Key) bool {
 	if len(prefix) > len(path) {
 		return false
 	}
 
 	return slices.Equal(prefix, path[:len(prefix)])
 }
+
+// opNode represents a node in the operation tree, tracking deletion status
+type opNode struct {
+	key       Key
+	children  map[Key]*opNode
+	isDeleted bool
+}
+
+// newOpNode creates a new operation tree node with the given key.
+func newOpNode(key Key) *opNode {
+	return &opNode{
+		key:      key,
+		children: make(map[Key]*opNode),
+	}
+}
+
+// collect returns an ordered subset of the input operation log based on the deletion status tracked in this operation
+// tree. The original chronological order of operations is preserved. The status of each node in the path (not just the
+// operation's target) determines whether the operation is included in the result.
+func (n *opNode) collect(oplog OperationLog) OperationLog {
+	var result OperationLog
+
+	for _, op := range oplog {
+		path := op.Path()
+
+		// Handle root node case
+		if len(path) == 0 {
+			if n.isDeleted {
+				if deleteOp, ok := op.(DeleteOperation); ok {
+					result = append(result, deleteOp)
+				}
+			} else {
+				result = append(result, op)
+			}
+			continue
+		}
```

Comment:
Seems like there should be a way to merge this with the `Check "target" node` code below.

Reply:
Merged!